### PR TITLE
Feature/download trajectory

### PR DESCRIPTION
**Pull request by interim17**
Problem
=======
Users need to be able to download their trajectories.

Closes [#295](https://github.com/simularium/simularium-website/issues/295)

Solution
========
Basic UI button and functionality is here to download networked files from our data bucket, as well as local files. 

Ability to download local files is redundant now, but will become relevant in future as we capture user changes in evolving Simularium file type. Recent updates to viewer have provided a method for getting blobs from JSON files, matching that functionality for binary files is forthcoming.

with @meganrm and @toloudis 

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Locate download button in top right of header.
2. Load a networked trajectory from examples, download it.
3. Load a local file, maybe the one you just downloaded, download it.

### CODE REVIEW

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        if (!fileIsLoaded()) {
+            return "";
+        }
+        if (isNetworkSimFileInterface(simulariumFile)) {
+            return `${DATA_BUCKET_URL}/trajectory/${simulariumFile.name}`;
+        } else {
+            const data: ISimulariumFile = simulariumFile.data;
+            // won't work if they loaded a binary file
+            // TODO: need to wait for simulariumFile.data.getAsBlob() to be created on viewer;
+            const blob = data.getAsBlob();
```
[the remark this answers is not in the document]
Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:36]: done

**Comment by meganrm** [comment]
I tried to verify this but I couldn't get any of the remote trajectories to load

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            const data: ISimulariumFile = simulariumFile.data;
+            // won't work if they loaded a binary file
+            const blob = data.getAsBlob();
+            return URL.createObjectURL(blob);
+        }
+    };
+
+    const downloadFile = async (fileName: string): Promise<void> => {
+        const downloadLink = document.createElement("a");
+        setIsDownloading(true);
```
Comment (meganrm) [inline src/components/DownloadTrajectoryMenu/index.tsx:43]: have you seen the button switch to "Downloading..."? Because none of these are written as async functions, it seems like the time between `setIsDownloading(true)` and `setIsDownloading(false)` is always going to be imperceivable

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        document.body.removeChild(downloadLink);
+        URL.revokeObjectURL(downloadLink.href);
+        setIsDownloading(false);
+    };
+
+    const onClick = () => {
+        if (!fileIsLoaded()) {
+            return;
+        }
+        downloadFile(simulariumFile.name);
```
Comment (meganrm) [inline src/components/DownloadTrajectoryMenu/index.tsx:58]:
```suggestion
        setIsDownloading(true);
        await downloadFile(simulariumFile.name);
        setIsDownloading(false);
```

and remove them from the other function

Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:58]: actually scratch this, I don't think it's ever possible to see the change in the tooltip, and that seems like the only reason for the `isDownloading` flag? When I click 'download' I never see it change it "Downloading..."

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    return (
+        <div className={styles.container}>
+            <Tooltip
+                placement="left"
+                title={
+                    isDownloading
+                        ? "Downloading..."
+                        : simulariumFile.name
+                        ? "Download trajectory"
+                        : "Load a trajectory to perform this action"
```
Comment (meganrm) [inline src/components/DownloadTrajectoryMenu/index.tsx:70]:
```suggestion
                    simulariumFile.name
                        ? "Download trajectory"
                        : "Load a trajectory to perform this action"
```

Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:70]: if you were going to keep the `isDowloading` I would suggest you change this to a "getTooltip" function with if else statements for readability

File: src/components/Icons/index.tsx [hunk 5]
```diff
@@ -21,12 +22,13 @@ export const Loading = <LoadingOutlined style={{ fontSize: 40 }} spin />;
 export const Play = <CaretRightOutlined />;
 export const Pause = <PauseOutlined />;
 export const UploadFile = <ImportOutlined />;
-export const DownArrow = <DownOutlined />;
+export const DownArrow = <DownOutlined size={32} />;
```
Comment (meganrm) [inline src/components/Icons/index.tsx:25]: why was this change needed?

File: src/components/Icons/index.tsx [hunk 6]
```diff
@@ -21,12 +22,13 @@ export const Loading = <LoadingOutlined style={{ fontSize: 40 }} spin />;
...
 export const CaretRight = <CaretRightOutlined />;
 export const GoBack = <ArrowLeftOutlined />;
 export const Reset = <HomeOutlined />;
 export const ZoomIn = <PlusOutlined />;
 export const ZoomOut = <MinusOutlined />;
+export const Download = <DownloadOutlined size={32} />;
```
Comment (meganrm) [inline src/components/Icons/index.tsx:31]: this seems redundant with `DownArrow`?

Comment (interim17) [reply src/components/Icons/index.tsx:31]: They are different icons, the download one has an arrow pointing into a tray.

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 4 above)
Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:70]: Ok, taking your suggestion.

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 3 above)
Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: @meganrm when you download a networked file do you get anything at all, or do you get a file that doesn't load properly in the viewer?

Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: I wasn't able to load a networked file. I just loaded a local file and then downloaded it again, which did work

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:35]: getAsBlob works for either binary or JSON, so you don't need the other comment either

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 3 above)
Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: unfortunately this seems to be a bug with the viewer version 3.6.0. I revered it back (and removed the `getAsBlob`) and the networked files loaded. @toloudis or @ascibisz do you know why this is? Here are the changes in the new release: https://github.com/simularium/simularium-viewer/releases/tag/v3.6.0, and here is the error I'm getting 
``` 
[Debug] [netconnection] – "WS Connection Request Sent: " – "ws://staging-node1-agentviz-backend.cellexplore.net:9002/" (vendor.d1aa0b7eabb278739d2c.js, line 54046)

[Error] WebSocket connection to 'ws://staging-node1-agentviz-backend.cellexplore.net:9002/' failed: The network connection was lost.
```

Comment (toloudis) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: I suspect it's this one: https://github.com/simularium/simularium-viewer/pull/316 and maybe we need a patch release?

Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: yeah that's it. this issue wasn't showing up on my machine because i manually fixed the ws/wss issue in node modules yesterday after working with Alli.

Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: yeah, I tested it with main and it works. I also got this other error
```
Uncaught TypeError: Cannot read properties of undefined (reading 'sendWebSocketRequest')
```
So I added [a PR to fix it](https://github.com/simularium/simularium-viewer/pull/317), but I'm not sure why the websocketClient is undefined

Comment (meganrm) [reply src/components/DownloadTrajectoryMenu/index.tsx:55]: This has now all been merged in and I published a new version. You should update the viewer to 6.0.1 and then it should be good to go

**Review by meganrm — APPROVED** [review]
Approve, but don't forget to remove the binary comment first

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        if (!fileIsLoaded()) {
+            return;
+        }
+        downloadFile(simulariumFile.name);
+    };
+
+    return (
+        <div className={styles.container}>
+            <Tooltip
+                placement="left"
```
Comment (frasercl) [inline src/components/DownloadTrajectoryMenu/index.tsx:60]: Tooltip covers the buttons to the left. Should it go below instead?

File: src/components/DownloadTrajectoryMenu/index.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    const downloadFile = async (fileName: string): Promise<void> => {
+        const downloadLink = document.createElement("a");
+        downloadLink.download = fileName;
+        downloadLink.style.display = "none";
+        downloadLink.href = getHref();
+        document.body.appendChild(downloadLink);
+        downloadLink.click();
+        document.body.removeChild(downloadLink);
+        URL.revokeObjectURL(downloadLink.href);
+    };
```
Comment (frasercl) [inline src/components/DownloadTrajectoryMenu/index.tsx:48]: This function doesn't need to be `async` unless it contains an `await` (I'm pretty sure).

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 7 above)
Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:60]: Reset it to "bottomLeft" which seems to obscure less

File: src/components/DownloadTrajectoryMenu/index.tsx (hunk 8 above)
Comment (interim17) [reply src/components/DownloadTrajectoryMenu/index.tsx:48]: You are correct

**Review by frasercl — APPROVED** [review]
